azurerm/resource_arm_private_link_service.go

```diff
@@ -0,0 +1,649 @@
+										ValidateFunc: validate.NoEmptyStrings,
+									},
+									"description": {
+										Type:         schema.TypeString,
+										Optional:     true,
+										ValidateFunc: validate.NoEmptyStrings,
+									},
+									"status": {
+										Type:         schema.TypeString,
+										Optional:     true,
+										ValidateFunc: validate.NoEmptyStrings,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+
+			"alias": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"network_interface_ids": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"type": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"tags": tags.Schema(),
+		},
+	}
+}
+
+func resourceArmPrivateLinkServiceCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).network.PrivateLinkServiceClient
+	ctx := meta.(*ArmClient).StopContext
+
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	if features.ShouldResourcesBeImported() && d.IsNewResource() {
+		resp, err := client.Get(ctx, resourceGroup, name, "")
+		if err != nil {
+			if !utils.ResponseWasNotFound(resp.Response) {
+				return fmt.Errorf("Error checking for present of existing Private Link Service %q (Resource Group %q): %+v", name, resourceGroup, err)
+			}
+		}
+		if !utils.ResponseWasNotFound(resp.Response) {
+			return tf.ImportAsExistsError("azurerm_private_link_service", *resp.ID)
+		}
+	}
+
+	location := azure.NormalizeLocation(d.Get("location").(string))
+	autoApproval := d.Get("auto_approval_subscription_ids").([]interface{})
+	fqdns := d.Get("fqdns").([]interface{})
+	ipConfigurations := d.Get("nat_ip_configuration").([]interface{})
+	loadBalancerFrontendIpConfigurations := d.Get("load_balancer_frontend_ip_configuration_ids").([]interface{})
+	privateEndpointConnections := d.Get("private_endpoint_connection").([]interface{})
+	visibility := d.Get("visibility_subscription_ids").([]interface{})
+	t := d.Get("tags").(map[string]interface{})
+
+	parameters := network.PrivateLinkService{
+		Location: utils.String(location),
+		PrivateLinkServiceProperties: &network.PrivateLinkServiceProperties{
+			AutoApproval:                         expandArmPrivateLinkServicePrivateLinkServicePropertiesAutoApproval(autoApproval),
+			Fqdns:                                utils.ExpandStringSlice(fqdns),
+			IPConfigurations:                     expandArmPrivateLinkServicePrivateLinkServiceIPConfiguration(ipConfigurations),
+			LoadBalancerFrontendIPConfigurations: expandArmPrivateLinkServiceFrontendIPConfiguration(loadBalancerFrontendIpConfigurations),
+			PrivateEndpointConnections:           expandArmPrivateLinkServicePrivateEndpointConnection(privateEndpointConnections),
+			Visibility:                           expandArmPrivateLinkServicePrivateLinkServicePropertiesVisibility(visibility),
```

Comment:
could we set this inline and remove the expan function
```suggestion
			Visibility:                           network.PrivateLinkServicePropertiesVisibility{
					Subscriptions: & utils.expandStringSlice(d.Get("visibility_subscription_ids").([]interface{})),
```

Reply:
Fixed. Same think here kept it in the expand function but switch to using `utils.ExpandStringSlice` because it made the code look weird.